nasa/harmony
# Harmony 1272 smart work item timeouts

## Jira Issue ID
HARMONY-1272

## Description
Uses a simple heuristic to retry stalled running work items that are relatively old based on previously successful work items for the same job/service.

## Local Test Steps
1. Add the following code to tasks/service-runner/app/workers/pull-worker.ts after line 170 to simulate stuck workers
```
// TODO remove this when done testing
if (Math.random() < 0.333) {
  return;
}
```
2. Build the service runner image using `npm run build` in the tasks/service-runner directory.
3. Run `bin/restart-services`
4. Run a query with multiple work items and watch the work items in the workflow-ui. You should see them fail about 1/3 of the time and get retried rather quickly (way less than two hours). The retry interval should increase as longer successful work items come in, and it should be about 1.5 times the duration of the slowest successful work item.

## PR Acceptance Checklist
